### ISSUE TITLE
Not compatible with HTML checkbox lists

### ISSUE THREAD

**Issue by ZettaScript** [seanmonstar/warp#733]
**Version**
warp: 0.2.5

**Platform**
Linux 5.4.72-1-lts, ArchLinux, x86_64  
Rust 1.49.0-nightly

**Description**
Lists encoded in `application/x-www-form-urlencoded` POST requests are not correctly deserialized.

For example, this HTML form:
```
<form method="post">
    <input type="checkbox" name="foo" value="1"/>
    <input type="checkbox" name="foo" value="2"/>
    <input type="checkbox" name="foo" value="3"/>
    <input type="submit"/>
</form>
```

And this Rust code:
```
warp::serve(
    warp::post().and(
        warp::body::form::<FooQuery>()
            .map(|q: FooQuery| format!("Request: {:?}", q))
    )
    .or(warp::get().map(handle_index)) // handle_index just serves HTML page
).run(address).await;

#[derive(Clone, Debug, Deserialize)]
struct FooQuery {
    foo: Vec<u32>,
}
```

The request's body can be `foo=2&foo=3` for example. It should be interpreted as `FooQuery { foo: vec![2, 3] }`, but there is an error: `Request body deserialize error: invalid type: string "2", expected a sequence`.

PHP needs field names to end with `[]` to interpret them as a list. I tried that but there is also an error.

**Comment by jxs** [comment]
Hi, warp uses `serde-urlencoded` for that feature, see https://github.com/nox/serde_urlencoded/issues/52#issuecomment-483138575 and https://github.com/nox/serde_urlencoded/issues/75#issuecomment-648257888

**Comment by Gelox** [comment]
@jxs Does that mean that it is not possible to parse vectors from a GET URL unless you choose to parse the raw string on your own?

**Comment by jxs** [comment]
@Gelox yeah, I was just going to answer there, [query filter]( https://github.com/seanmonstar/warp/blob/master/src/filters/query.rs#L21) also uses `serde_urlencoded`

**Comment by Gelox** [comment]
@jxs Is there interest in updating this so vectors are supported via get requests? I understand `serde` is not interested in changing this but it seems like a modern HTTP library would want to handle vector support in some way.
Have similar issues been closed with the decision to not support this?

**Comment by jxs** [comment]
@Gelox yeah, similar issues have been raised: https://github.com/seanmonstar/warp/issues/677, https://github.com/seanmonstar/warp/issues/22 and on the discord channel a couple times. So maybe we could consider to switch to `serde_qs` since `serde_urlencoded` is not interested in offering those features, what do you think @seanmonstar ?

**Comment by Gelox** [comment]
As far as I can tell `serde_qs` does not have the same list syntax for query strings that is usually the case for web frameworks, at least from the two sources I posted in the other issue [stackoverflow](https://stackoverflow.com/questions/3061273/send-an-array-with-an-http-get) and [wiki](https://en.wikipedia.org/wiki/Query_string#Web_forms).
`serde_qs` wants an array to be indexed like so
`/get_path?list[0]=a&list[1]=b`
while the (not standardized) syntax in the links is either
`/get_path?list=a&list=b`
or
`/get_path?list[]=a&list[]=b`

of course the syntax in `serde_qs` does make it clearer what the ordering will be in a vector but it might be a concern that it requires users to use a different syntax than they are used to for specifying lists.
Of course `serde_qs` is better than nothing, but perhaps some additions can be made to the implementations? I will ask them what their thoughts on this are.

**Comment by kaj** [comment]
@Gelox :
> @jxs Does that mean that it is not possible to parse vectors from a GET URL unless you choose to parse the raw string on your own?

You can get the query as a `Vec<(String, String)>` and go from there on your own.   This way,  `/get_path?list=a&list=b` will get the handler called with `[("list", "a"), ("list", "b")]`.  I use this in my project rphotos:

https://github.com/kaj/rphotos/blob/master/src/server/search.rs#L121-L136

**Comment by Kerndog73** [comment]
A somewhat related issue is that using single checkboxes is a little bit unpleasant. If the checkbox is checked, it will be present and set to `"on"` (unless the `value` attribute is set to something else). If the checkbox is unchecked, it won't be present. This means you need this to deserialize the parameter into `Option<String>` rather than `bool`. This is probably more a serde issue but I thought I'd bring it up anyway.

**Comment by MarcAntoine-Arnaud** [comment]
Axum has just changed `serde_urlencoded` by `serde_qs`
https://github.com/tokio-rs/axum/issues/504

**Comment by jrandall** [comment]
There now also appears to be a crate `serde_html_form` (https://github.com/jplatte/serde_html_form) which is a fork of `serde_urlencoded` but which supports deserializing URLs with repeated form fields into a sequence (e.g. `Vec<String>`).

**Comment by jplatte** [comment]
> Axum has just changed `serde_urlencoded` by `serde_qs` [tokio-rs/axum#504](https://github.com/tokio-rs/axum/issues/504)

This is not true, axum still uses `serde_urlencoded`. There are however `Query` and `Form` extractors in `axum-extra` that use `serde_html_form`.

**Comment by 42triangles** [comment]
Unless I am mistaken somewhere, it appears that `body::form` isn't compatible with checkboxes *period*, as it doesn't handle a single checkbox correctly either as far as I can tell - [`serde_urlencoded` expects a `true` or `false` for booleans](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e8197899d550e4bb7d72584be7783c11), even though [web browsers either don't send it (if it wasn't checked) or send `on` (or some user defined value; if it was checked)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e8197899d550e4bb7d72584be7783c11).

**Comment by jplatte** [comment]
@42triangles Well, it's debatable what values should produce a boolean `true`, but for omission producing `false` that's only possible with a `#[serde(default)]` attribute given serde's data model. Without that attribute¹, missing fields are a deserialization error regardless of the deserializer in use.

¹ or the field being `Option<_>`-typed, which acts like the attribute was given except in combination with some other attributes
